### PR TITLE
Complete calculator project

### PR DESCRIPTION
**Pull request by reid-andy**
Completed Calculator project plus non-AI challenges (though not ALL the suggested extra calculations). 

Things got a little borked between my repo and Microsoft's vs-tutorial-samples repo for a minute. I untangled them, but not expertly. 

The history feature was a fun piece to work on. I tried implementing a custom type/class (are those the same thing?) as was recommended in the math game project I turned in. 

I'm sure Codacy is going to have a lot to say about this project, but I'm interested to know what you think about where I put functions and classes for the challenges. 

Thank you for everything you do!

### CODE REVIEW

File: CalculatorLibrary/CalculatorLibrary.cs [hunk 1]
```diff
@@ -0,0 +1,176 @@
+﻿using Newtonsoft.Json;
+using System.Diagnostics;
```
Comment (nwdorian) [inline CalculatorLibrary/CalculatorLibrary.cs:2]: 🟠**Unnecessary using directives**

➡️ Remove anything that's not being used. It makes your code less readable and maintainable. 

💡In Visual Studio the compiler will tell you when code isn't being used by slightly fading it. Here's a short video on how to make Visual Studio [automatically remove unused using statements](https://www.youtube.com/watch?v=CC5nohNp7F8)

Image

```suggestion
﻿using Newtonsoft.Json;
```

File: CalculatorLibrary/CalculatorLibrary.cs [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+
+
+
```
Comment (nwdorian) [inline CalculatorLibrary/CalculatorLibrary.cs:5]: 🟡 **Whitespace**

❓ There are a couple of unnecessary blank lines. Use only one line of whitespace for code separation.

➡️ Remove the extra whitespace to keep the code visually consistent and aligned with standard C# formatting conventions.

File: CalculatorLibrary/CalculatorLibrary.cs [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+namespace CalculatorLibrary
+{
```
Comment (nwdorian) [inline CalculatorLibrary/CalculatorLibrary.cs:7]: 🟠 **Namespaces**  

❓As per the academy's [code-conventions](https://thecsharpacademy.com/article/58/code-conventions), use file-scoped namespaces.  

➡️ From C# 10 we are able to remove the namespace block and save an indentation level, which optimizes the use of space and improves readability.

💡 If you're using Visual Studio 2022, here's a short article to setup [file-scoped namespaces by default](https://darthpedro.net/2024/01/29/configure-visual-studio-to-use-file-scoped-namespaces-on-new-classes/).


```suggestion
namespace CalculatorLibrary;
```

File: CalculatorLibrary/CalculatorLibrary.cs [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+            double result = double.NaN; // Default value is "not-a-number" which we use if an operation, such as division, could result in an error.
+            writer.WriteStartObject();
+            writer.WritePropertyName("Operand1");
+            writer.WriteValue(num1);
+            writer.WritePropertyName("Operand2");
+            writer.WriteValue(num2);
+            writer.WritePropertyName("Operation");
+
+            // Use a switch statement to do the math. TEST PR
+            switch (op)
```
Comment (nwdorian) [inline CalculatorLibrary/CalculatorLibrary.cs:34]: 🟡 **Code comments**

❓ It's fine to add comments during development if it's going to help you with pseudo-code or to remember some specific logic. When you finish a project, always go back through the code for a polish pass and clean up unnecessary comments.  

💭 Remember, our code should be self-documenting - with good naming and structure we shouldn't need any comments at all.

File: CalculatorLibrary/CalculatorLibrary.cs [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+
+            }
+
+            return null;
+        }
+
+    }
+
+    public class Calculation
+    {
```
Comment (nwdorian) [inline CalculatorLibrary/CalculatorLibrary.cs:130]: 🟠 **One class per file**

❓ Each top-level class, struct, record, or interface should be defined in its own file.

➡️ This practices keeps files organized and improves maintainability and discoverability. It also aligns with standard .NET conventions and matches the expectations of IDE tools and code analyzers.

File: CalculatorLibrary/CalculatorLibrary.cs [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+            get; set;
+        }
+        public string? operation
+        {
+            get; set;
+        }
+        public int calculationId
+        {
+            get; set;
+        }
```
Comment (nwdorian) [inline CalculatorLibrary/CalculatorLibrary.cs:150]: 🟠 **Formatting**

❓ While properties are technically methods under the hood, they should follow standard C# property formatting conventions, which place the entire auto-property on a single line and use PascalCase naming.

💡 Using multi-line braces here adds unnecessary vertical space, makes the class harder to scan, and breaks typical .NET style expectations.

➡️ Rewrite the properties using the standard single-line auto-property syntax and make them PascalCase following [.NET naming conventions](https://learn.microsoft.com/en-us/dotnet/standard/design-guidelines/capitalization-conventions#capitalization-rules-for-identifiers).


```suggestion
        public double Num1 { get; set; }
        public double Num2 { get; set; }
        public double Result { get; set; }
        public string? Operation { get; set; }
        public int CalculationId { get; set; }
```

File: csharp/Calculator/Program.cs [hunk 7]
```diff
@@ -0,0 +1,100 @@
+using CalculatorLibrary;
+using System;
+using System.Collections.Generic;
+namespace CalculatorProgram
```
Comment (nwdorian) [inline csharp/Calculator/Program.cs:4]: 🟡 **Whitespace**

➡️ There should be a blank line here to separate using declarations from the namespace.

```suggestion
using System.Collections.Generic;

namespace CalculatorProgram
```

**Review by nwdorian — APPROVED** [review]
Hey @reid-andy 👋,

Great work on your **Calculator** project submission! 🎉

I have finished the code review and left some feedback below. ☺️

Feel free to reach out if you have questions or if you'd like to collaborate on any of these points.✌️

---

🟢 **Requirements**

⭐ You have completed all of the project requirements!

---

🏅 **Extra challenges**

✅ Count the amount of times the calculator was used
✅ List of latest calculations
✅ Use previous results for new calculations
✅ Add extra operations
💪 Very nice to see you implemented extra challenges as it speeds up the learning process! 

---

🟠 **Submission Structure**

❓ As per the academy's [Code Reviews](https://thecsharpacademy.com/article/52/code-reviews) page, I was expecting your project submission to be contained in a folder with the format: **ProjectName.GitHubUserName**

👉 For example, `Calculator.reid-andy` with all application related files and folders inside it. The submission should be in self-contained top-level folder with correct naming. Take a look at this structure for example:

📁 `CodeReviews.Console.Calculator` ⬅️ Forked repo
 ┣ 📄 `.codacy.yml`
 ┣ 📄 `.gitignore`
 ┣ 📁 `Calculator.reid-andy` ⬅️ Top-level folder
 ┃  ┗ 📄 `Calculator.sln`
 ┃  ┗ 📄 `README.md`
 ┃  ┗ 📄 `LICENSE`
 ┃  ┗ 📁 `Calculator.Console`
 ┃  ┗ 📁 `Calculator.Library`

💡The easiest way to tackle this problem is to create a new folder for your submission right after cloning the fork. This will make it easier for code-reviewers to review your code in the future. It will also make sense when you get to the projects which contain other students code.

---

🟠 **Multiple solution files**

❓ There are multiple solution files in the submission. You should only have one solution file per codebase.

➡️ When adding new projects you should right-click the solution file in Visual Studio 2022 and add the project to the solution. I'm just assuming you were creating a new app which resulted in another solution file.

---

🟠 **Bug - selecting a result from history doesn't work**

❓ After entering an entry number to use a result from history, the prompt to enter two new numbers still appears. 

➡️ Make sure to thoroughly test your application before submission. Make sure the happy path works, then cover any edge cases you can think of. 

Image

---

🟡 **UX improvements**

➡️ Whenever you are designing console applications, make sure to clear the console after each operation or it gets cluttered pretty quickly.

🛠️ Add `Console.Clear()` at start or end of the while loop.

---

💭 You can take care of issues in your own repo if you want, but since the project will be approved you don't have to push the changes or create a new PR.

✅ I will go ahead and approve the project submission. Keep up the good work on the next ones! 😄

Best regards,  
@nwdorian
